Add support for loading nodes, fields, and paragraphs.

damontgomery opened the pull request:
Nodes can be loaded by type / title / language.
Fields can be loaded.
Entity references are loaded.
Special handling is used for fields with "nothing" values, they are checked to see if they are empty.

Special handling for:
- Entity references which includes nodes, terms, media
- Long text
- Date (not yet ranges)
- Links
- Translations
- Paragraph types in fields
- Fields on paragraphs

Example:
```
Scenario: Blog post exists
  When I examine the "blog_post" node with title "My post"
  Then entity field "field_simple_example" should contain "test"
  And entity field "field_is_a_thing" should contain "nothing"
  # Example entity reference
  And entity field "field_authors" should contain "John Smith"
  And paragraph field "field_paragraphs" should be of type "my_paragraph_type"

Scenario: Paragraph fields
  When I examine the "page" node with title "My page"
  And I examine paragraph "5" on the "field_paragraphs" field
  Then entity field "field_content" should contain "Example"

Scenario: Paragraph fields in English
  When I examine the "page" node with title "My page" in "en"
  And I examine paragraph "5" on the "field_paragraphs" field
  Then entity field "field_content" should contain "Example"
```

damontgomery committed:
Add support for loading nodes and fields.

Nodes can be loaded by type / title.
Fields can be loaded.
Entity references are loaded.
Special handling is used for fields with "nothing" values, they are checked to see if they are empty.

Example:
```
Scenario: Blog post exists
  Given I examine the "blog_post" node with title "My post"
  Then entity property "title" should be "My post"
  And entity field "field_simple_example" should contain "test"
  And entity field "field_is_a_thing" should contain "nothing"
  # Example entity reference
  And entity field "field_authors" should contain "John Smith"
```

damontgomery committed:
Exclude field name since the step already states it and it's harder to get.

damontgomery commented:
Going to work on this a little more.

damontgomery committed:
Use one method for properties and fields.

Since properties and fields are accessed the same way.

damontgomery committed:
Handle field values better.

After testing on more examples.

damontgomery committed:
Remove termreference since all entity references are shared in D8.

damontgomery committed:
Handle different types of entity references.

Fix docs around input parameters
Add docs for $currentEntity

damontgomery committed:
Update functions since we pass field as an object.

damontgomery committed:
Add support for long text.

damontgomery committed:
Add very basic support for date comparisons.

damontgomery committed:
Handle media file references.

The default seems to be `name` rather than `title` which is used for nodes.

damontgomery committed:
Handle paragraph fields

Paragraphs don't have titles, but we can check the type. A new pattern has been added.
Paragraphs are entity references, so we throw an error when the general field pattern is being used since we aren't checking for value / title.

damontgomery commented:
I missed out on the shared node extension. :(

damontgomery committed:
Update node lookup method.

This was needed before, but I forgot to export it.

damontgomery committed:
Add paragraph examine step.

Can be used like this
```
When I examine the "page" node with title "My page"
And I examine paragraph "5" on the "field_paragraphs" field
Then entity field "field_content" should contain "Example"
```

damontgomery commented:
Added support for loading paragraphs.

Commit:
Link URL field behat context testing

Commit:
Updating composer.json

Commit:
Revert "Updating composer.json"

This reverts commit d4f733b695ae1c2022e4bfbfd54e558d199a20f1.

damontgomery committed:
Add support for testing translations, including paragraph translations.

damontgomery commented:
I've included the link work in #35 and also added support for translations.

becw commented on src/Palantirnet/PalantirBehatExtension/Context/EntityDataContext.php:
```diff
@@ -340,43 +399,66 @@ public function assertNotEntityFieldValue($field, $value)
     }//end assertNotEntityFieldValue()
 
+    /**
+     * Verify that a field contains a value.
```
I think this line of the comment is not accurate?

damontgomery replied:
Thanks, I'll update this.

becw commented on src/Palantirnet/PalantirBehatExtension/Context/EntityDataContext.php:
```diff
@@ -459,133 +531,97 @@ public function assertEntityFieldValueImage($field, $value)
+
+                switch ($entity->getEntityTypeId()) {
+                    case 'node':
+                        $title = $entity->title->value;
```
Can you use `$entity->label()` here instead of relying on local logic for finding entity labels?

Also, I don't think Paragraph entities will come up here, since they are referenced by the "Entity Reference Revisions" field type, rather than the "Entity Reference" field type.

becw replied:
... ah, except for the `assertEntityFieldValueEntityReferenceRevisions` methods below! I see.

damontgomery replied:
Thanks, I should be able to use `$entity->label()` for everything but paragraphs (their label isn't useful). I can update this and it should be a little clearer.

becw commented on src/Palantirnet/PalantirBehatExtension/Context/SharedDrupalContext.php:
```diff
@@ -36,21 +36,36 @@ class SharedDrupalContext extends RawDrupalContext
-            return node_load($nid);
-        } else if (empty($entities['node']) === false && count($entities['node']) > 1) {
+            return $node;
+        } else if (count($entities) > 1) {
```
I think this condition won't ever be hit, even when it should be -- the logic above seems to load and return the first nid in the array, regardless of whether the array has more than one result. Is this accurate?

damontgomery replied:
Line `51` has `if (count($entities) === 1) {`, so the node will only be loaded if there is a single match. I've definitely hit this many times during tested and used a different node for testing. The logic is from the D7 version and I think it makes sense since there is no good way to know which node to use.

I think some of the confusion is that `$nid = array_shift($entities);` is used to get the first element. I think the keys are node IDs, so `$entities[0]` may not work but `array_keys($entities)[0]` would if that is clearer.

becw reviewed (commented):
I've done a visual review of the code here -- is there any additional way you'd like me to test this?

BTW, the paragraph handling you added is pretty cool!

damontgomery committed:
Address review feedback.

- Update paragraph assert method description.
- Replace switch statement with generalized `$entity->label()`.
- Add comment around node loading logic.

damontgomery commented:
@becw,

Thanks for the feedback. I've made the updates you requested. This has been working for us on our project. Please let me know if you want to see some of it in detail on our project.

I'm good with the code as a first version.

If it looks good, do you want to merge it?
